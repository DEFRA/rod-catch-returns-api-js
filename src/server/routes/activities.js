import {
  Activity,
  Catch,
  River,
  SmallCatch,
  SmallCatchCount,
  Submission
} from '../../entities/index.js'
import {
  createActivitySchema,
  updateActivitySchema
} from '../../schemas/activities.schema.js'
import {
  extractRiverId,
  extractSubmissionId
} from '../../utils/entity-utils.js'
import { handleNotFound, handleServerError } from '../../utils/server-utils.js'
import { StatusCodes } from 'http-status-codes'
import logger from '../../utils/logger-utils.js'
import { mapActivityToResponse } from '../../mappers/activities.mapper.js'
import { mapCatchToResponse } from '../../mappers/catches.mapper.js'
import { mapRiverToResponse } from '../../mappers/river.mapper.js'
import { mapSmallCatchToResponse } from '../../mappers/small-catches.mapper.js'
import { mapSubmissionToResponse } from '../../mappers/submission.mapper.js'
import { sequelize } from '../../services/database.service.js'

const BASE_ACTIVITIES_URL = '/activities/{activityId}'

export default [
  {
    method: 'POST',
    path: '/activities',
    options: {
      /**
       * Create an activity for a submission in the database
       *
       * @param {import('@hapi/hapi').Request request - The Hapi request object
       *     @param {string} request.payload.submission - The submission id prefixed with submissions/
       *     @param {string} request.payload.daysFishedWithMandatoryRelease - The number of days fished during the mandatory release period
       *     @param {string} request.payload.daysFishedOther - The number of days fished at other times during the season
       *     @param {string} request.payload.river - The submission id prefixed with rivers/
       * @param {import('@hapi/hapi').ResponseToolkit} h - The Hapi response toolkit
       * @returns {Promise<import('@hapi/hapi').ResponseObject>} - A response containing the target {@link Activity}
       */
      handler: async (request, h) => {
        try {
          const {
            submission,
            daysFishedWithMandatoryRelease,
            daysFishedOther,
            river
          } = request.payload

          const activityData = {
            submission_id: extractSubmissionId(submission),
            river_id: extractRiverId(river),
            daysFishedWithMandatoryRelease,
            daysFishedOther,
            version: new Date()
          }

          logger.info('Creating activity with details', activityData)

          const createdActivity = await Activity.create(activityData)

          const response = mapActivityToResponse(createdActivity.toJSON())

          return h.response(response).code(StatusCodes.CREATED)
        } catch (error) {
          return handleServerError('Error creating activity', error, h)
        }
      },
      validate: {
        payload: createActivitySchema,
        options: { entity: 'Activity' }
      },
      description: 'Create an activity for a submission in the database',
      notes: 'Create an activity for a submission in the database',
      tags: ['api', 'activities']
    }
  },
  {
    method: 'GET',
    path: `${BASE_ACTIVITIES_URL}/river`,
    options: {
      /**
       * Retrieve the river associated with an activity
       *
       * @param {import('@hapi/hapi').Request request - The Hapi request object
       *     @param {string} request.params.activityId - The activity id
       * @param {import('@hapi/hapi').ResponseToolkit} h - The Hapi response toolkit
       * @returns {Promise<import('@hapi/hapi').ResponseObject>} - A response containing the target {@link River}
       */
      handler: async (request, h) => {
        try {
          const activityId = request.params.activityId
          const activity = await Activity.findOne({
            where: { id: activityId },
            include: [
              {
                model: River,
                required: true
              }
            ]
          })

          if (!activity) {
            return handleNotFound(
              'Activity not found or has no associated river',
              h
            )
          }

          const mappedRiver = mapRiverToResponse(activity.River.toJSON())

          return h.response(mappedRiver).code(StatusCodes.OK)
        } catch (error) {
          return handleServerError(
            'Error fetching river for activity',
            error,
            h
          )
        }
      },
      description:
        'Retrieve the river associated with an activity in the database',
      notes: 'Retrieve river associated with an activity in the database',
      tags: ['api', 'activities']
    }
  },
  {
    method: 'GET',
    path: `${BASE_ACTIVITIES_URL}/smallCatches`,
    options: {
      /**
       * Retrieve the small catches associated with an activity
       *
       * @param {import('@hapi/hapi').Request request - The Hapi request object
       *     @param {string} request.params.activityId - The activity id
       * @param {import('@hapi/hapi').ResponseToolkit} h - The Hapi response toolkit
       * @returns {Promise<import('@hapi/hapi').ResponseObject>} - A response containing the target {@link SmallCatch}
       */
      handler: async (request, h) => {
        try {
          const activityId = request.params.activityId
          const activityWithCatches = await Activity.findOne({
            where: { id: activityId },
            include: [
              {
                model: SmallCatch,
                include: [
                  {
                    association: 'counts'
                  }
                ]
              }
            ]
          })

          if (!activityWithCatches) {
            return handleNotFound(
              `Small catches not found for ${activityId}`,
              h
            )
          }

          const mappedSmallCatches = (
            activityWithCatches.SmallCatches || []
          ).map((smallCatch) => mapSmallCatchToResponse(smallCatch))

          return h
            .response({
              _embedded: {
                smallCatches: mappedSmallCatches
              }
            })
            .code(StatusCodes.OK)
        } catch (error) {
          return handleServerError('Error fetching small catches', error, h)
        }
      },
      description:
        'Retrieve the small catches associated with an activity in the database',
      notes: 'Retrieve small catches with an activity in the database',
      tags: ['api', 'activities']
    }
  },
  {
    method: 'GET',
    path: `${BASE_ACTIVITIES_URL}/catches`,
    options: {
      /**
       * Retrieve the catches (salmon and large sea trout) associated with an activity
       *
       * @param {import('@hapi/hapi').Request request - The Hapi request object
       *     @param {string} request.params.activityId - The activity id
       * @param {import('@hapi/hapi').ResponseToolkit} h - The Hapi response toolkit
       * @returns {Promise<import('@hapi/hapi').ResponseObject>} - A response containing the target {@link Catch}
       */
      handler: async (request, h) => {
        try {
          const activityId = request.params.activityId
          const activityWithCatches = await Activity.findOne({
            where: { id: activityId },
            include: [
              {
                model: Catch
              }
            ]
          })

          if (!activityWithCatches) {
            return handleNotFound(`Catches not found for ${activityId}`, h)
          }

          const mappedCatches = (activityWithCatches.Catches || []).map(
            (catchEntity) => mapCatchToResponse(catchEntity)
          )

          return h
            .response({
              _embedded: {
                catches: mappedCatches
              }
            })
            .code(StatusCodes.OK)
        } catch (error) {
          return handleServerError('Error fetching catches', error, h)
        }
      },
      description:
        'Retrieve the catches (salmon and large sea trout) associated with an activity in the database',
      notes:
        'Retrieve catches (salmon and large sea trout) with an activity in the database',
      tags: ['api', 'activities']
    }
  },
  {
    method: 'GET',
    path: BASE_ACTIVITIES_URL,
    options: {
      /**
       * Retrieve an activity by its ID
       *
       * @param {import('@hapi/hapi').Request request - The Hapi request object
       *     @param {string} request.params.activityId - The activity id
       * @param {import('@hapi/hapi').ResponseToolkit} h - The Hapi response toolkit
       * @returns {Promise<import('@hapi/hapi').ResponseObject>} - A response containing the target {@link Activity}
       */
      handler: async (request, h) => {
        try {
          const activityId = request.params.activityId
          const activity = await Activity.findOne({
            where: { id: activityId }
          })

          if (!activity) {
            return handleNotFound(`Activity not found for ID: ${activityId}`, h)
          }

          const mappedActivity = mapActivityToResponse(activity.toJSON())

          return h.response(mappedActivity).code(StatusCodes.OK)
        } catch (error) {
          return handleServerError('Error fetching activity by ID', error, h)
        }
      },
      description: 'Retrieve an activity by its ID',
      notes: 'Retrieve an activity from the database by its ID',
      tags: ['api', 'activities']
    }
  },
  {
    method: 'DELETE',
    path: BASE_ACTIVITIES_URL,
    options: {
      /**
       * Delete an activity by ID
       *
       * @param {import('@hapi/hapi').Request request - The Hapi request object
       *     @param {string} request.params.activityId - The activity id
       * @param {import('@hapi/hapi').ResponseToolkit} h - The Hapi response toolkit
       * @returns {Promise<import('@hapi/hapi').ResponseObject>} - A response indicating success or failure
       */
      handler: async (request, h) => {
        const activityId = request.params.activityId

        // Begin transaction for atomic operation
        const transaction = await sequelize.transaction()

        try {
          // Find IDs of all SmallCatch records associated with the Activity
          logger.info(
            'Deleting activity with id:%s and related records',
            activityId
          )
          const smallCatchIds = (
            await SmallCatch.findAll({
              attributes: ['id'],
              where: { activity_id: activityId },
              transaction
            })
          )?.map((smallCatch) => smallCatch.id)

          // Delete associated SmallCatchCount
          await SmallCatchCount.destroy({
            where: { small_catch_id: smallCatchIds },
            transaction
          })

          // Delete associated SmallCatches
          await SmallCatch.destroy({
            where: { activity_id: activityId },
            transaction
          })

          // Delete associated Catches
          await Catch.destroy({
            where: { activity_id: activityId },
            transaction
          })

          // Delete the Activity
          const deletedCount = await Activity.destroy({
            where: { id: activityId },
            transaction
          })

          if (deletedCount === 0) {
            await transaction.rollback()
            return handleNotFound(`Activity with ID ${activityId} not found`, h)
          }

          // Commit transaction
          await transaction.commit()

          logger.info(
            'Deleted activity with id: %s and related records',
            activityId
          )
          return h.response().code(StatusCodes.NO_CONTENT)
        } catch (error) {
          await transaction.rollback()
          return handleServerError('Error deleting activity', error, h)
        }
      },
      description: 'Delete an activity by ID',
      notes: 'Deletes an activity from the database by its ID',
      tags: ['api', 'activities']
    }
  },
  {
    method: 'PATCH',
    path: BASE_ACTIVITIES_URL,
    options: {
      /**
       * Update a activities in the database using the activities ID
       *
       * @param {import('@hapi/hapi').Request request - The Hapi request object
       *     @param {string} request.params.activityId - The ID of the activity to update
       * @param {import('@hapi/hapi').ResponseToolkit} h - The Hapi response toolkit
       * @returns {Promise<import('@hapi/hapi').ResponseObject>} - A response containing the target {@link Activity}
       */
      handler: async (request, h) => {
        const { activityId } = request.params
        const { daysFishedWithMandatoryRelease, daysFishedOther, river } =
          request.payload

        try {
          const activity = await Activity.findByPk(activityId)

          if (!activity) {
            return handleNotFound(`Activity not found for ${activityId}`, h)
          }

          const activityData = {
            river_id: river ? extractRiverId(river) : undefined,
            daysFishedWithMandatoryRelease,
            daysFishedOther,
            version: new Date()
          }

          logger.info(
            `Updating activity ${activityId} with details`,
            activityData
          )

          // if a value is undefined, it is not updated by Sequelize
          const updatedActivity = await activity.update(activityData)

          const mappedActivity = mapActivityToResponse(updatedActivity.toJSON())

          return h.response(mappedActivity).code(StatusCodes.OK)
        } catch (error) {
          return handleServerError('Error updating activity', error, h)
        }
      },
      validate: {
        payload: updateActivitySchema,
        options: { entity: 'Activity' }
      },
      description: 'Update an activity',
      notes: 'Update an activity',
      tags: ['api', 'activities']
    }
  },
  {
    method: 'GET',
    path: `${BASE_ACTIVITIES_URL}/submission`,
    options: {
      /**
       * Retrieve the submission associated with an activity
       *
       * @param {import('@hapi/hapi').Request request - The Hapi request object
       *     @param {string} request.params.activityId - The activity id
       * @param {import('@hapi/hapi').ResponseToolkit} h - The Hapi response toolkit
       * @returns {Promise<import('@hapi/hapi').ResponseObject>} - A response containing the target {@link Submission}
       */
      handler: async (request, h) => {
        try {
          const activityId = request.params.activityId
          const activity = await Activity.findOne({
            where: { id: activityId },
            include: [
              {
                model: Submission,
                required: true
              }
            ]
          })

          if (!activity) {
            return handleNotFound(
              'Activity not found or has no associated submission',
              h
            )
          }

          const mappedSubmission = mapSubmissionToResponse(
            activity.Submission.toJSON()
          )

          return h.response(mappedSubmission).code(StatusCodes.OK)
        } catch (error) {
          return handleServerError(
            'Error fetching submission for activity',
            error,
            h
          )
        }
      },
      description:
        'Retrieve the submission associated with an activity in the database',
      notes: 'Retrieve submission associated with an activity in the database',
      tags: ['api', 'activities']
    }
  }
]
