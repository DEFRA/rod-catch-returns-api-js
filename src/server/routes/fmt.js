import { authorize, signIn } from '../../services/oidc-service.js'

export default [
  {
    path: '/oidc/authorize',
    method: 'GET',
    handler: authorize,
    options: { auth: false }
  },
  {
    path: '/oidc/callback',
    method: 'POST',
    handler: signIn,
    options: { auth: false }
  }
]
