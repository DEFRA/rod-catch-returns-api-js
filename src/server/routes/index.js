import Catchments from './catchments.js'
import Licences from './licences.js'
import Region from './regions.js'
import Rivers from './rivers.js'
import Species from './species.js'

import Static from './static.js'

export const apiPrefixRoutes = [].concat(Licences, Region, Rivers, Catchments, Species)

export const rootRoutes = [].concat(Static)
