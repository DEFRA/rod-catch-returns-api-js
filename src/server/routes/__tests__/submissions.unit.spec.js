import {
  getMockResponseToolkit,
  getServerDetails
} from '../../../test-utils/server-test-utils.js'
import {
  handleNotFound,
  handleServerError
} from '../../../utils/server-utils.js'
import { Submission } from '../../../entities/index.js'
import { createActivity as createActivityCRM } from '@defra-fish/dynamics-lib'
import { getCreateActivityResponse } from '../../../test-utils/test-data.js'
import logger from '../../../utils/logger-utils.js'
import routes from '../submissions.js'

jest.mock('../../../entities/index.js')
jest.mock('../../../utils/logger-utils.js')
jest.mock('../../../utils/server-utils.js')

const [
  {
    options: { handler: postSubmissionHandler }
  },
  {
    options: { handler: getSubmissionByContactIdAndSeasonHandler }
  },
  {
    options: { handler: getActivitiesBySubmissionIdHandler }
  },
  {
    options: { handler: getSubmissionByIdHandler }
  },
  {
    options: { handler: patchSubmissionByIdHandler }
  }
] = routes

const NOT_FOUND_SYMBOL = Symbol('NOT_FOUND')
const SERVER_ERROR_SYMBOL = Symbol('SERVER_ERROR')

handleNotFound.mockReturnValue(NOT_FOUND_SYMBOL)
handleServerError.mockReturnValue(SERVER_ERROR_SYMBOL)

describe('submissions.unit', () => {
  const getFoundSubmission = () => ({
    toJSON: jest.fn().mockReturnValue({
      id: '1',
      contactId: 'contact-identifier-111',
      season: '2024',
      status: 'COMPLETE',
      source: 'WEB',
      version: '2024-10-10T13:13:11.000Z',
      reportingExclude: false,
      createdAt: '2024-10-10T13:13:11.000Z',
      updatedAt: '2024-10-10T13:13:11.000Z'
    })
  })

  describe('POST /submissions', () => {
    const getSubmissionRequest = () =>
      getServerDetails({
        payload: {
          contactId: 'contact-identifier-111',
          season: '2024',
          status: 'INCOMPLETE',
          source: 'WEB'
        }
      })

    const getCreatedSubmission = () => ({
      toJSON: jest.fn().mockReturnValue({
        id: '1',
        contactId: 'contact-identifier-111',
        season: '2024',
        status: 'INCOMPLETE',
        source: 'WEB',
        version: '2024-10-10T13:13:11.000Z',
        reportingExclude: false,
        createdAt: '2024-10-10T13:13:11.000Z',
        updatedAt: '2024-10-10T13:13:11.000Z'
      })
    })

    afterEach(() => {
      jest.clearAllMocks()
    })

    it('should return a 201 status code if the submission is created successfully', async () => {
      Submission.create.mockResolvedValueOnce(getCreatedSubmission())
      createActivityCRM.mockResolvedValueOnce(getCreateActivityResponse())

      const result = await postSubmissionHandler(
        getSubmissionRequest(),
        getMockResponseToolkit()
      )

      expect(result.statusCode).toBe(201)
    })

    it('should return the created submission in the response body', async () => {
      createActivityCRM.mockResolvedValueOnce(getCreateActivityResponse())
      Submission.create.mockResolvedValueOnce(getCreatedSubmission())

      const result = await postSubmissionHandler(
        getSubmissionRequest(),
        getMockResponseToolkit()
      )

      expect(result.payload).toStrictEqual({
        contactId: 'contact-identifier-111',
        createdAt: '2024-10-10T13:13:11.000Z',
        id: '1',
        reportingExclude: false,
        season: '2024',
        source: 'WEB',
        status: 'INCOMPLETE',
        updatedAt: '2024-10-10T13:13:11.000Z',
        version: '2024-10-10T13:13:11.000Z',
        _links: {
          activities: {
            href: 'http://localhost:3000/api/submissions/1/activities'
          },
          self: {
            href: 'http://localhost:3000/api/submissions/1'
          },
          submission: {
            href: 'http://localhost:3000/api/submissions/1'
          }
        }
      })
    })

    it('should call handleServerError if submission creation fails', async () => {
      const error = new Error('Database error')
      Submission.create.mockRejectedValueOnce(error)

      const h = getMockResponseToolkit()

      await postSubmissionHandler(getSubmissionRequest(), h)

      expect(handleServerError).toHaveBeenCalledWith(
        'Error creating submission',
        error,
        h
      )
    })

    it('should return a error response if an error occurs while creating submission', async () => {
      const error = new Error('Database error')
      Submission.create.mockRejectedValueOnce(error)

      const result = await postSubmissionHandler(
        getSubmissionRequest(),
        getMockResponseToolkit()
      )

      expect(result).toBe(SERVER_ERROR_SYMBOL)
    })

    it('should log an error but still return 201 when the call to create an activity in CRM returns an ErrorMessage', async () => {
      Submission.create.mockResolvedValueOnce(getCreatedSubmission())
      createActivityCRM.mockResolvedValue({
        '@odata.context':
          'https://dynamics.com/api/data/v9.1/defra_CreateRCRActivityResponse',
        RCRActivityId: null,
        ReturnStatus: 'error',
        SuccessMessage: '',
        ErrorMessage: 'Failed to create activity'
      })

      const result = await postSubmissionHandler(
        getSubmissionRequest(),
        getMockResponseToolkit()
      )

      expect(result.statusCode).toBe(201)
      expect(logger.error).toHaveBeenCalledWith(
        'failed to create activity in CRM for contact-identifier-111',
        'Failed to create activity'
      )
    })

    it('should call handleServerError when the call to create an activity in CRM returns an error', async () => {
      Submission.create.mockResolvedValueOnce(getCreatedSubmission())
      const error = new Error('CRM')
      createActivityCRM.mockRejectedValueOnce(error)

      const h = getMockResponseToolkit()
      await postSubmissionHandler(getSubmissionRequest(), h)

      expect(handleServerError).toHaveBeenCalledWith(
        'Error creating submission',
        error,
        h
      )
    })

    it('should return an error response when the call to create an activity in CRM returns an error', async () => {
      Submission.create.mockResolvedValueOnce(getCreatedSubmission())
      const error = new Error('CRM')
      createActivityCRM.mockRejectedValueOnce(error)

      const result = await postSubmissionHandler(
        getSubmissionRequest(),
        getMockResponseToolkit()
      )

      expect(result).toBe(SERVER_ERROR_SYMBOL)
    })
  })

  describe('GET /submissions/search/getByContactIdAndSeason', () => {
    const getSubmissionRequest = () =>
      getServerDetails({
        query: {
          contact_id: 'contact-identifier-111',
          season: '2024'
        }
      })

    afterEach(() => {
      jest.clearAllMocks()
    })

    it('should return a 200 status code if the submission is found', async () => {
      Submission.findOne.mockResolvedValueOnce(getFoundSubmission())

      const result = await getSubmissionByContactIdAndSeasonHandler(
        getSubmissionRequest(),
        getMockResponseToolkit()
      )

      expect(result.statusCode).toBe(200)
    })

    it('should return the found submission in the response body', async () => {
      Submission.findOne.mockResolvedValueOnce(getFoundSubmission())

      const result = await getSubmissionByContactIdAndSeasonHandler(
        getSubmissionRequest(),
        getMockResponseToolkit()
      )

      expect(result.payload).toMatchSnapshot()
    })

    it('should call handleNotFound if the submission is not found', async () => {
      Submission.findOne.mockResolvedValueOnce(null)
      const h = getMockResponseToolkit()

      await getSubmissionByContactIdAndSeasonHandler(getSubmissionRequest(), h)

      expect(handleNotFound).toHaveBeenCalledWith(
        'Submission not found for contact-identifier-111 and 2024',
        h
      )
    })

    it('should return a not found response if the submission is not found', async () => {
      Submission.findOne.mockResolvedValueOnce(null)

      const result = await getSubmissionByContactIdAndSeasonHandler(
        getSubmissionRequest(),
        getMockResponseToolkit()
      )

      expect(result).toBe(NOT_FOUND_SYMBOL)
    })

    it('should call handleServerError if fetching a submission fails', async () => {
      const error = new Error('Database error')
      Submission.findOne.mockRejectedValueOnce(error)
      const h = getMockResponseToolkit()

      await getSubmissionByContactIdAndSeasonHandler(getSubmissionRequest(), h)

      expect(handleServerError).toHaveBeenCalledWith(
        'Error finding submission',
        error,
        h
      )
    })

    it('should return an error message if an error occurs while fetching submission', async () => {
      const error = new Error('Database error')
      Submission.findOne.mockRejectedValueOnce(error)

      const result = await getSubmissionByContactIdAndSeasonHandler(
        getSubmissionRequest(),
        getMockResponseToolkit()
      )

      expect(result).toBe(SERVER_ERROR_SYMBOL)
    })
  })

  describe('GET /submissions/{submissionId}', () => {
    const getSubmissionRequest = () =>
      getServerDetails({
        params: {
          submissionId: '1'
        }
      })

    afterEach(() => {
      jest.clearAllMocks()
    })

    it('should return a 200 status code if the submission is found', async () => {
      Submission.findOne.mockResolvedValueOnce(getFoundSubmission())

      const result = await getSubmissionByIdHandler(
        getSubmissionRequest(),
        getMockResponseToolkit()
      )

      expect(result.statusCode).toBe(200)
    })

    it('should return the found submission in the response body', async () => {
      Submission.findOne.mockResolvedValueOnce(getFoundSubmission())

      const result = await getSubmissionByIdHandler(
        getSubmissionRequest(),
        getMockResponseToolkit()
      )

      expect(result.payload).toStrictEqual({
        contactId: 'contact-identifier-111',
        createdAt: '2024-10-10T13:13:11.000Z',
        id: '1',
        reportingExclude: false,
        season: '2024',
        source: 'WEB',
        status: 'COMPLETE',
        updatedAt: '2024-10-10T13:13:11.000Z',
        version: '2024-10-10T13:13:11.000Z',
        _links: {
          activities: {
            href: 'http://localhost:3000/api/submissions/1/activities'
          },
          self: {
            href: 'http://localhost:3000/api/submissions/1'
          },
          submission: {
            href: 'http://localhost:3000/api/submissions/1'
          }
        }
      })
    })

    it('should call handleNotFound if the submission is not found', async () => {
      Submission.findOne.mockResolvedValueOnce(null)
      const h = getMockResponseToolkit()

      await getSubmissionByIdHandler(getSubmissionRequest(), h)

      expect(handleNotFound).toHaveBeenCalledWith('Submission not found 1', h)
    })

    it('should call a not found response if the submission is not found', async () => {
      Submission.findOne.mockResolvedValueOnce(null)

      const result = await getSubmissionByIdHandler(
        getSubmissionRequest(),
        getMockResponseToolkit()
      )

      expect(result).toBe(NOT_FOUND_SYMBOL)
    })

    it('should call handleServerError if fetching submission fails', async () => {
      const error = new Error('Database error')
      Submission.findOne.mockRejectedValueOnce(error)
      const h = getMockResponseToolkit()

      await getSubmissionByIdHandler(getSubmissionRequest(), h)

      expect(handleServerError).toHaveBeenCalledWith(
        'Error finding submission',
        error,
        h
      )
    })

    it('should return an error response  if an error occurs while fetching submission', async () => {
      const error = new Error('Database error')
      Submission.findOne.mockRejectedValueOnce(error)

      const result = await getSubmissionByIdHandler(
        getSubmissionRequest(),
        getMockResponseToolkit()
      )

      expect(result).toBe(SERVER_ERROR_SYMBOL)
    })
  })

  describe('GET /submissions/{submissionId}/activities', () => {
    const getFoundSubmissionWithActivities = (activities = []) => ({
      id: '1',
      contactId: 'contact-identifier-111',
      season: '2024',
      status: 'COMPLETE',
      source: 'WEB',
      version: '2024-10-10T13:13:11.000Z',
      Activities: activities,
      createdAt: '2024-10-10T13:13:11.000Z',
      updatedAt: '2024-10-10T13:13:11.000Z'
    })

    const getActivityMock = () => ({
      toJSON: jest.fn().mockReturnValue({
        id: '1',
        daysFishedWithMandatoryRelease: 1,
        daysFishedOther: 2,
        createdAt: '2024-10-10T13:13:11.000Z',
        updatedAt: '2024-10-10T13:13:11.000Z',
        version: '2024-10-10T13:13:11.000Z'
      })
    })

    const getActivitiesRequest = () =>
      getServerDetails({
        params: {
          submissionId: '1'
        }
      })

    afterEach(() => {
      jest.clearAllMocks()
    })

    it('should return 200 with activities if they exist for the submission', async () => {
      const foundSubmissionWithActivities = getFoundSubmissionWithActivities([
        getActivityMock()
      ])
      Submission.findOne.mockResolvedValueOnce(foundSubmissionWithActivities)

      const result = await getActivitiesBySubmissionIdHandler(
        getActivitiesRequest(),
        getMockResponseToolkit()
      )

      expect(result.payload).toStrictEqual({
        _embedded: {
          activities: [
            {
              id: '1',
              daysFishedWithMandatoryRelease: 1,
              daysFishedOther: 2,
              createdAt: '2024-10-10T13:13:11.000Z',
              updatedAt: '2024-10-10T13:13:11.000Z',
              version: '2024-10-10T13:13:11.000Z',
              _links: {
                self: {
                  href: 'http://localhost:3000/api/activities/1'
                },
                activity: {
                  href: 'http://localhost:3000/api/activities/1'
                },
                submission: {
                  href: 'http://localhost:3000/api/activities/1/submission'
                },
                catches: {
                  href: 'http://localhost:3000/api/activities/1/catches'
                },
                river: {
                  href: 'http://localhost:3000/api/activities/1/river'
                },
                smallCatches: {
                  href: 'http://localhost:3000/api/activities/1/smallCatches'
                }
              }
            }
          ]
        }
      })
      expect(result.statusCode).toBe(200)
    })

    it('should return 200 with an empty activities array if the submission exists but no activities are found', async () => {
      Submission.findOne.mockResolvedValueOnce(
        getFoundSubmissionWithActivities()
      )

      const result = await getActivitiesBySubmissionIdHandler(
        getActivitiesRequest(),
        getMockResponseToolkit()
      )

      expect(result.payload).toStrictEqual({ _embedded: { activities: [] } })
      expect(result.statusCode).toBe(200)
    })

    it('should call handleNotFound if the submission does not exist', async () => {
      Submission.findOne.mockResolvedValueOnce(null)
      const h = getMockResponseToolkit()

      await getActivitiesBySubmissionIdHandler(getActivitiesRequest(), h)

      expect(handleNotFound).toHaveBeenCalledWith(
        'Activities not found for submission with id 1',
        h
      )
    })

    it('should return a not found response if the submission does not exist', async () => {
      Submission.findOne.mockResolvedValueOnce(null)

      const result = await getActivitiesBySubmissionIdHandler(
        getActivitiesRequest(),
        getMockResponseToolkit()
      )

      expect(result).toBe(NOT_FOUND_SYMBOL)
    })

    it('should log an error if fetching submission with activities fails', async () => {
      const error = new Error('Database error')
      Submission.findOne.mockRejectedValueOnce(error)
      const h = getMockResponseToolkit()

      await getActivitiesBySubmissionIdHandler(getActivitiesRequest(), h)

      expect(handleServerError).toHaveBeenCalledWith(
        'Error finding activities for submission',
        error,
        h
      )
    })

    it('should an error response if an error occurs while fetching submission with activities', async () => {
      const error = new Error('Database error')
      Submission.findOne.mockRejectedValueOnce(error)

      const result = await getActivitiesBySubmissionIdHandler(
        getActivitiesRequest(),
        getMockResponseToolkit()
      )

      expect(result).toBe(SERVER_ERROR_SYMBOL)
    })
  })

  describe('PATCH /submissions/{submissionId}', () => {
    const getSubmissionRequest = (payload) =>
      getServerDetails({
        params: {
          submissionId: '1'
        },
        payload
      })

    const getFoundSubmission = () => ({
      update: jest.fn().mockResolvedValue({
        toJSON: jest.fn().mockReturnValue({
          id: '1',
          contactId: 'contact-identifier-111',
          season: '2024',
          status: 'COMPLETE',
          source: 'WEB',
          version: '2024-10-10T13:13:11.000Z',
          reportingExclude: false,
          createdAt: '2024-10-10T13:13:11.000Z',
          updatedAt: '2024-10-10T13:13:11.000Z'
        })
      })
    })

    afterEach(() => {
      jest.clearAllMocks()
    })

    it('should return a 200 status code if the submission is updated successfully', async () => {
      Submission.findByPk.mockResolvedValueOnce(getFoundSubmission())

      const result = await patchSubmissionByIdHandler(
        getSubmissionRequest({ status: 'COMPLETE' }),
        getMockResponseToolkit()
      )

      expect(result.statusCode).toBe(200)
    })

    it('should call update with the "status"', async () => {
      const foundSubmission = getFoundSubmission()
      Submission.findByPk.mockResolvedValueOnce(foundSubmission)

      await patchSubmissionByIdHandler(
        getSubmissionRequest({ status: 'COMPLETE' }),
        getMockResponseToolkit()
      )

      expect(foundSubmission.update).toHaveBeenCalledWith({
        status: 'COMPLETE',
        reportingExclude: undefined,
        version: expect.any(Date)
      })
    })

    it('should call update with "reportingExclude"', async () => {
      const foundSubmission = getFoundSubmission()
      Submission.findByPk.mockResolvedValueOnce(foundSubmission)

      await patchSubmissionByIdHandler(
        getSubmissionRequest({ reportingExclude: true }),
        getMockResponseToolkit()
      )

      expect(foundSubmission.update).toHaveBeenCalledWith({
        status: undefined,
        reportingExclude: true,
        version: expect.any(Date)
      })
    })

    it('should return the updated submission in the response body', async () => {
      Submission.findByPk.mockResolvedValueOnce(getFoundSubmission())

      const result = await patchSubmissionByIdHandler(
        getSubmissionRequest({ status: 'COMPLETE' }),
        getMockResponseToolkit()
      )

      expect(result.payload).toMatchSnapshot()
    })

    it('should call handleNotFound if the submission is not found', async () => {
      Submission.findByPk.mockResolvedValueOnce(null)
      const h = getMockResponseToolkit()

      await patchSubmissionByIdHandler(
        getSubmissionRequest({ status: 'COMPLETE' }),
        h
      )

      expect(handleNotFound).toHaveBeenCalledWith(
        'Submission not found for 1',
        h
      )
    })

    it('should return a not found response if the submission is not found', async () => {
      Submission.findByPk.mockResolvedValueOnce(null)

      const result = await patchSubmissionByIdHandler(
        getSubmissionRequest({ status: 'COMPLETE' }),
        getMockResponseToolkit()
      )

      expect(result).toBe(NOT_FOUND_SYMBOL)
    })

    it('should call handleServerError if updating the submission fails', async () => {
      const error = new Error('Database error')
      Submission.findByPk.mockRejectedValueOnce(error)
      const h = getMockResponseToolkit()

      await patchSubmissionByIdHandler(
        getSubmissionRequest({ status: 'COMPLETE' }),
        h
      )

      expect(handleServerError).toHaveBeenCalledWith(
        'Error updating submission',
        error,
        h
      )
    })

    it('should return an error response if an error occurs while updating the submission', async () => {
      const error = new Error('Database error')
      Submission.findByPk.mockRejectedValueOnce(error)

      const result = await patchSubmissionByIdHandler(
        getSubmissionRequest({ status: 'COMPLETE' }),
        getMockResponseToolkit()
      )

      expect(result).toBe(SERVER_ERROR_SYMBOL)
    })
  })
})
