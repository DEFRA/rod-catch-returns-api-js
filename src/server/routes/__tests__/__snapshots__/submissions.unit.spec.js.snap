// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`submissions.unit GET /submissions/search/getByContactIdAndSeason should return the found submission in the response body 1`] = `
{
  "_links": {
    "activities": {
      "href": "http://localhost:3000/api/submissions/1/activities",
    },
    "self": {
      "href": "http://localhost:3000/api/submissions/1",
    },
    "submission": {
      "href": "http://localhost:3000/api/submissions/1",
    },
  },
  "contactId": "contact-identifier-111",
  "createdAt": "2024-10-10T13:13:11.000Z",
  "id": "1",
  "reportingExclude": false,
  "season": "2024",
  "source": "WEB",
  "status": "COMPLETE",
  "updatedAt": "2024-10-10T13:13:11.000Z",
  "version": "2024-10-10T13:13:11.000Z",
}
`;

exports[`submissions.unit PATCH /submissions/{submissionId} should return the updated submission in the response body 1`] = `
{
  "_links": {
    "activities": {
      "href": "http://localhost:3000/api/submissions/1/activities",
    },
    "self": {
      "href": "http://localhost:3000/api/submissions/1",
    },
    "submission": {
      "href": "http://localhost:3000/api/submissions/1",
    },
  },
  "contactId": "contact-identifier-111",
  "createdAt": "2024-10-10T13:13:11.000Z",
  "id": "1",
  "reportingExclude": false,
  "season": "2024",
  "source": "WEB",
  "status": "COMPLETE",
  "updatedAt": "2024-10-10T13:13:11.000Z",
  "version": "2024-10-10T13:13:11.000Z",
}
`;
